File: ratinmaze.java
import java.util.*;
public class ratinmaze{
    static void solve(int x, int y,int n,List<String>res,StringBuilder path,int [][] arr,List<Choice>choices){
        if(issolved(x,y,n)){
            res.add(path.toString());
            return;
        }
        for(Choice choice:choices){
            int newx=x+choice.dx;
            int newy=x+choice.dy;
            if(isvalid(newx , newy, n , arr)){
                arr[x][y]=0;
                path.append(choice.dname);
                solve(newx,newy,n,res,path,arr,choices);
                arr[x][y]=1;
                path.deleteCharAt(path.length()-1);
            } 
        }
    }
    public static boolean issolved(int x,int y,int n){
        return x==n-1 && y==n-1;
    }
    public static boolean isvalid(int x,int y,int n,int[][] arr){
        return x>=0 && x<n && y>=0 && y<n && arr[x][y]==1;
    }
    static class Choice{
        int dx;
        int dy;
        char dname;
        Choice(int dx,int dy,char dname){
            this.dx=dx;
            this.dy=dy;
            this.dname=dname;
        }
    }
    public static List<String>findpath(int[][] arr,int n){
        List<String>res=new ArrayList<>();
        StringBuilder path=new StringBuilder();
        List<Choice> choices=Arrays.asList(
            new Choice(-1,0,'u'),
            new Choice(1,0,'d'),
            new Choice(0,-1,'l'),
            new Choice(0,1,'r')
        );
        if(arr[0][0]==1){
            solve(0,0,n,res,path,arr,choices);
        }
        return res;
    }
    public static void mains(String [] args){
        int [][] maze={
            {1,0,0,0},
            {1,1,0,1},
            {0,1,0,0},
            {1,1,1,1}
        };
        int n=maze.length;
        List<String>paths=findpath(maze,n);
        if(paths.isEmpty()){
            System.out.println("No path found");
        }
        else{
            for(String path:paths){
                System.out.println(path);
            }
        }
    }
}

File: ratinmaze.py
import sys
from typing import List


class Choice:
    def __init__(self, dx: int, dy: int, name: str):
        self.dx = dx
        self.dy = dy
        self.name = name


def is_solved(x: int, y: int, n: int) -> bool:
    return x == n - 1 and y == n - 1


def is_valid(x: int, y: int, n: int, grid: List[List[int]]) -> bool:
    return 0 <= x < n and 0 <= y < n and grid[x][y] == 1


def dfs(x: int, y: int, n: int, path: List[str], grid: List[List[int]], choices: List[Choice], res: List[str]):
    if is_solved(x, y, n):
        res.append("".join(path))
        return
    for choice in choices:
        nx, ny = x + choice.dx, y + choice.dy
        if is_valid(nx, ny, n, grid):
            grid[x][y] = 0
            path.append(choice.name)
            dfs(nx, ny, n, path, grid, choices, res)
            path.pop()
            grid[x][y] = 1


def find_paths(grid: List[List[int]]) -> List[str]:
    n = len(grid)
    if n == 0 or grid[0][0] == 0:
        return []
    choices = [Choice(-1, 0, "u"), Choice(1, 0, "d"), Choice(0, -1, "l"), Choice(0, 1, "r")]
    res = []
    dfs(0, 0, n, [], grid, choices, res)
    return res


def main():
    data = list(map(int, sys.stdin.read().strip().split()))
    if not data:
        return
    n, *rest = data
    vals = rest[: n * n]
    grid = [vals[i * n : (i + 1) * n] for i in range(n)]
    paths = find_paths(grid)
    if not paths:
        print("No path found")
    else:
        for p in paths:
            print(p)


if __name__ == "__main__":
    main()
